### PR TITLE
MV-408: [MV-MIOpen] Benchmark & Port NLLLoss (add 2d and 5d)

### PR DESCRIPTION
**Pull request by hieule88**
* Added no reduction contiguous NLLLoss forward and backward operation and kernel.
* Added no reduction non-contiguous NLLLoss forward and backward operation and kernel.
* Added reduction NLLLoss forward and backward operation and kernel.
* Added driver test and gtest for NLLLoss.
* New API is guarded by MIOPEN_BETA_API macro.
* Compared to ROCm pytorch: [Link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7BB3BEF639-5EAE-4D05-81AD-6180091A454B%7D&file=NLLLoss_benchmark.xlsx&action=default&mobileredirect=true)

* Average over all cases:

Contiguous : 
|type    |Forward| Backward |
|--------|-------|-------|
|float16 | 5.27  | 4.09 |
|float32 | 4.03  | 2.82 | 
|bfloat16| 5.28  | 4.03 |

Non-Contiguous : 
|type    | Forward | Backward |
|--------|-------|-------------|
|float16 | 3.84  | 3.51 |
|float32 | 3.49  | 2.98 |
|bfloat16| 3.83  | 4.53 |

Reduction:
|type    |Forward| Backward |
|--------|-------|-------------|
|float16 | 1.66  | 2.00 |
|float32 | 1.71  | 1.97  |
|bfloat16| 1.66  | 2.03 |

This result does not include some instances where MIOpen significantly outperforms ROCm in cases with a large number of classes:

<details close>
<summary>Input size = [8192 52100] (N, C)</summary>
<br>

|    op_name   |  dtype   | contiguous | reduction | ignore index | direction | ROCm pytorch | MIOpen HIP |  Improvement  |
|:------------:|:--------:|:----------:|:---------:|:------------:|:---------:|:------------:|:-----------:|:------------:|
| NLLLoss      | float16  |    true    |    none   |     -100     |    fwd    |     28769    |    14437    | 1.99         |
| NLLLoss      | float16  |    true    |    none   |     -100     |    bwd    |     784551   |    14260    | 55.01        |
| NLLLoss      | float16  |    false   |    none   |     -100     |    fwd    |     2435181  |    14579    | 167.03       |
| NLLLoss      | float16  |    false   |    none   |     -100     |    bwd    |     3194068  |    14455    | 220.96       |
| NLLLoss      | float16  |    false   |    mean   |     -100     |    fwd    |     2759594  |    24642    | 111.98       |
| NLLLoss      | float16  |    false   |    mean   |     -100     |    bwd    |     3429560  |    12783    | 268.29       |
| NLLLoss      | float16  |    false   |    sum    |     -100     |    fwd    |     2760448  |    25104    | 109.96       |
| NLLLoss      | float16  |    false   |    sum    |     -100     |    bwd    |     3429032  |    12961    | 264.56       |
| NLLLoss      | float32  |    true    |    none   |     -100     |    fwd    |     33745    |    15574    | 2.16         |
| NLLLoss      | float32  |    true    |    none   |     -100     |    bwd    |     1184589  |    25957    | 45.63        |
| NLLLoss      | float32  |    false   |    none   |     -100     |    fwd    |     3299934  |    15396    | 214.33       |
| NLLLoss      | float32  |    false   |    none   |     -100     |    bwd    |     4476791  |    24890    | 179.86       |
| NLLLoss      | float32  |    false   |    mean   |     -100     |    fwd    |     3642497  |    34135    | 106.70       |
| NLLLoss      | float32  |    false   |    mean   |     -100     |    bwd    |     4738202  |    14152    | 334.80       |
| NLLLoss      | float32  |    false   |    sum    |     -100     |    fwd    |     3649286  |    25565    | 142.74       |
| NLLLoss      | float32  |    false   |    sum    |     -100     |    bwd    |     4741964  |    24428    | 194.12       |
| NLLLoss      | bfloat16 |    true    |    none   |     -100     |    fwd    |     28177    |    14276    | 1.97         |
| NLLLoss      | bfloat16 |    true    |    none   |     -100     |    bwd    |     784960   |    14738    | 53.26        |
| NLLLoss      | bfloat16 |    false   |    none   |     -100     |    fwd    |     2432283  |    14347    | 169.53       |
| NLLLoss      | bfloat16 |    false   |    none   |     -100     |    bwd    |     3191033  |    14632    | 218.08       |
| NLLLoss      | bfloat16 |    false   |    mean   |     -100     |    fwd    |     2758091  |    24783    | 111.28       |
| NLLLoss      | bfloat16 |    false   |    mean   |     -100     |    bwd    |     3430700  |    13085    | 262.18       |
| NLLLoss      | bfloat16 |    false   |    sum    |     -100     |    fwd    |     2758670  |    24641    | 111.95       |
| NLLLoss      | bfloat16 |    false   |    sum    |     -100     |    bwd    |     3435032  |    12711    | 270.24       |
</details>

<details close>
<summary>Input size = [20480 50000] (N, C)</summary>
<br>

|    op_name   |  dtype   | contiguous | reduction | ignore index | direction | ROCm pytorch |  MIOpen HIP |  Improvement  |
|:------------:|:--------:|:----------:|:---------:|:------------:|:---------:|:------------:|:-----------:|:-------------:|
| NLLLoss      | float16  |    true    |    none   |     -100     |    fwd    |    35521  	 |    20285	   |   1.75        |
| NLLLoss      | float16  |    true    |    none   |     -100     |    bwd    |    1852737	 |    21547	   |   85.98       |
| NLLLoss      | float16  |    false   |    none   |     -100     |    fwd    |    5804885	 |    19752	   |   293.88      |
| NLLLoss      | float16  |    false   |    none   |     -100     |    bwd    |    7600988	 |    21049	   |   361.10      |
| NLLLoss      | float16  |    false   |    mean   |     -100     |    fwd    |    6632306	 |    28800	   |   230.28      |
| NLLLoss      | float16  |    false   |    mean   |     -100     |    bwd    |    8166788	 |    19449	   |   419.90      |
| NLLLoss      | float16  |    false   |    sum    |     -100     |    fwd    |    6631614	 |    27697	   |   239.43      |
| NLLLoss      | float16  |    false   |    sum    |     -100     |    bwd    |    8173236	 |    19164	   |   426.48      |
| NLLLoss      | float32  |    true    |    none   |     -100     |    fwd    |    42704	 |    23501	   |   1.81        |
| NLLLoss      | float32  |    true    |    none   |     -100     |    bwd    |    2804726	 |    29812	   |   94.08       |
| NLLLoss      | float32  |    false   |    none   |     -100     |    fwd    |    7850643	 |    23217	   |   338.14      |
| NLLLoss      | float32  |    false   |    none   |     -100     |    bwd    |    10698810	 |    23591	   |   453.51      |
| NLLLoss      | float32  |    false   |    mean   |     -100     |    fwd    |    8752699	 |    37706	   |   232.13      |
| NLLLoss      | float32  |    false   |    mean   |     -100     |    bwd    |    11695705	 |    28764	   |   406.60      |
| NLLLoss      | float32  |    false   |    sum    |     -100     |    fwd    |    8711305	 |    32585	   |   267.34      |
| NLLLoss      | float32  |    false   |    sum    |     -100     |    bwd    |    11648068	 |    26008	   |   447.86      |
| NLLLoss      | bfloat16 |    true    |    none   |     -100     |    fwd    |    35152	 |    20267	   |   1.73        |
| NLLLoss      | bfloat16 |    true    |    none   |     -100     |    bwd    |    1854038	 |    21280	   |   87.12       |
| NLLLoss      | bfloat16 |    false   |    none   |     -100     |    fwd    |    5798155	 |    20355	   |   284.85      |
| NLLLoss      | bfloat16 |    false   |    none   |     -100     |    bwd    |    7598346	 |    20728	   |   366.57      |
| NLLLoss      | bfloat16 |    false   |    mean   |     -100     |    fwd    |    6649970	 |    27786	   |   239.32      |
| NLLLoss      | bfloat16 |    false   |    mean   |     -100     |    bwd    |    8159293	 |    18880	   |   432.16      |
| NLLLoss      | bfloat16 |    false   |    sum    |     -100     |    fwd    |    6624722	 |    28978	   |   228.61      |
| NLLLoss      | bfloat16 |    false   |    sum    |     -100     |    bwd    |    8167933	 |    20533	   |   397.79      |
</details>

### CODE REVIEW

File: driver/nllloss_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+}
+
+template <typename Tgpu, typename Tref>
+int NLLLossDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only Forward NLLLoss (Default=1)", "int");
+    inflags.AddInputFlag("input_dims",
+                         'D',
+                         "16,21,21,21,10",
+                         "The dimensional lengths of the input tensor",
```
Comment (et16kr) [inline driver/nllloss_driver.hpp]: It would be helpful to have an input sample for the dims.

File: driver/nllloss_driver.hpp [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+                         "string");
+    inflags.AddInputFlag("contiguous",
+                         'c',
+                         "1",
+                         "Is input tensor contiguous? (Default=1 for contiguous tensor)",
+                         "int");
+
+    inflags.AddInputFlag("iter", 'i', "10", "Number of Iterations (Default=10)", "int");
+    inflags.AddInputFlag("verify", 'V', "1", "Verify (Default=1)", "int");
+    inflags.AddInputFlag("time", 't', "1", "Time (Default=0)", "int");
```
Comment (et16kr) [inline driver/nllloss_driver.hpp]: The default for time is shown as 0, but in reality, it is 1. Please make it consistent.

**Review by et16kr — COMMENTED** [review]
The bfloat16 test is failing. Please check it.
```
MIOPEN_TEST_FLOAT_ARG=--bfloat16 ./test_nllloss

[ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
/home/MIOpen/test/gtest/nllloss.hpp:239: Failure
Value of: error < threshold * 10
  Actual: false
Expected: true
Error output beyond tolerance Error:0.13130428906372638,  Thresholdx10: 0.078125

[  FAILED  ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12, where GetParam() =  input:{8192,52100} weight_mode:1 ignore_index:-100 divisor:1 Contiguous:0 (2132 ms)
```

File: driver/nllloss_driver.hpp (hunk 1 above)
Comment (hieule88) [reply driver/nllloss_driver.hpp]: fixed it

File: driver/nllloss_driver.hpp (hunk 2 above)
Comment (hieule88) [reply driver/nllloss_driver.hpp]: fixed it

**Comment by hieule88** [comment]
> The bfloat16 test is failing. Please check it.
> 
> ```
> MIOPEN_TEST_FLOAT_ARG=--bfloat16 ./test_nllloss
> 
> [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
> /home/MIOpen/test/gtest/nllloss.hpp:239: Failure
> Value of: error < threshold * 10
>   Actual: false
> Expected: true
> Error output beyond tolerance Error:0.13130428906372638,  Thresholdx10: 0.078125
> 
> [  FAILED  ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12, where GetParam() =  input:{8192,52100} weight_mode:1 ignore_index:-100 divisor:1 Contiguous:0 (2132 ms)
> ```

My code passes the Driver test for this input even when I change the Tref to bfloat. The unit test passes for the values of the workspaces but still fails for the value of the reduce sum. I tried many ways, but it still doesn't work. The error remains at 0.13 (while the threshold is 0.07). Can we tolerate this?

**Comment by et16kr** [comment]
> > The bfloat16 test is failing. Please check it.
> > ```
> > MIOPEN_TEST_FLOAT_ARG=--bfloat16 ./test_nllloss
> > 
> > [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
> > /home/MIOpen/test/gtest/nllloss.hpp:239: Failure
> > Value of: error < threshold * 10
> >   Actual: false
> > Expected: true
> > Error output beyond tolerance Error:0.13130428906372638,  Thresholdx10: 0.078125
> > 
> > [  FAILED  ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12, where GetParam() =  input:{8192,52100} weight_mode:1 ignore_index:-100 divisor:1 Contiguous:0 (2132 ms)
> > ```
> 
> My code passes the Driver test for this input even when I change the Tref to bfloat. The unit test passes for the values of the workspaces but still fails for the value of the reduce sum. I tried many ways, but it still doesn't work. The error remains at 0.13 (while the threshold is 0.07). Can we tolerate this?

Forgive me for being late.

In the kernel, bfloat16 is converted to float for calculation.
Try the same as kernel.

```
root@a06c94249c27:/home/MIOpen/build# git diff
diff --git a/test/cpu_nllloss.hpp b/test/cpu_nllloss.hpp
index 87dff7338..7b5e258a9 100644
--- a/test/cpu_nllloss.hpp
+++ b/test/cpu_nllloss.hpp
@@ -68,6 +68,7 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
     auto I_tv = get_inner_expanded_tv_5d(input.desc);
     auto T_tv = get_inner_expanded_tv_4d(target.desc);
     auto W_tv = get_inner_expanded_tv_1d(weight.desc);
+    float d   = !std::isnan(divisor) ? static_cast<float>(divisor) : static_cast<float>(1.0f);

     for(size_t i = 0; i < numel; i++)
     {
@@ -84,11 +85,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
         }
         else
         {
-            T w = weight[weight_index];
-
-            T input_value = input[input_index];
-            T d           = !std::isnan(divisor) ? static_cast<T>(divisor) : static_cast<T>(1.0f);
-            workspace[i]  = (-w * input_value) / d;
+            float w = static_cast<float>(weight[weight_index]);
+            float input_value = static_cast<float>(input[input_index]);
+            workspace[i]  = static_cast<T>((-w * input_value) / d);
         }
     }

@@ -101,9 +100,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
     {
         for(int i = 0; i < _size; i += local_size)
         {
-            T shared[local_size];
+            float shared[local_size];
             par_ford(local_size)([&](size_t j) {
-                shared[j] = i + j < _size ? workspace[offset_a + i + j] : static_cast<T>(0.0f);
+               shared[j] = i + j < _size ? static_cast<float>(workspace[offset_a + i + j]) : static_cast<float>(0.0f);
             });
             for(int offset = local_size / 2; offset > 0; offset >>= 1)
                 ford(local_size)([&](size_t j) {
@@ -111,9 +110,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
                         shared[j] += shared[j + offset];
                 });
             if(_size <= local_size)
-                output[0] = shared[0];
+                output[0] = static_cast<T>(shared[0]);
             else
-                workspace[offset_b + i / local_size] = shared[0];
+                workspace[offset_b + i / local_size] = static_cast<T>(shared[0]);
         }
         std::swap(offset_a, offset_b);
         _size = (_size + local_size - 1) / local_size;
```


```
# MIOPEN_TEST_FLOAT_ARG=--bfloat16 bin/test_nllloss

[----------] 13 tests from NLLLossTestSet/NLLLossTestBFloat16
[ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/0
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '"/root/.cache/miopen/3.1.0.56a22e6bf"' is: 0x794c7630 'OVERLAYFS_SUPER_MAGIC'
MIOpen(HIP): Info [KernDb] database not present

...

[ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/11
[       OK ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/11 (0 ms)
[ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
[       OK ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12 (53164 ms)
[----------] 13 tests from NLLLossTestSet/NLLLossTestBFloat16 (55032 ms total)
```

File: test/gtest/nllloss.hpp [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+        auto out_dim     = divisor == 0.f ? target_dim : std::vector<size_t>{1};
+        auto out_strides = GetStrides(out_dim, true);
+        output           = tensor<T>{out_dim, out_strides};
+        std::fill(output.begin(), output.end(), std::numeric_limits<T>::quiet_NaN());
+
+        ref_output = tensor<T>{out_dim, out_strides};
+        std::fill(ref_output.begin(), ref_output.end(), std::numeric_limits<T>::quiet_NaN());
+
+        std::vector<size_t> workspace_lengths;
```
Comment (et16kr) [inline test/gtest/nllloss.hpp]: This vector appears to be unused.

Comment (hieule88) [reply test/gtest/nllloss.hpp]: removed it

**Comment by hieule88** [comment]
> > > The bfloat16 test is failing. Please check it.
> > > ```
> > > MIOPEN_TEST_FLOAT_ARG=--bfloat16 ./test_nllloss
> > > 
> > > [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
> > > /home/MIOpen/test/gtest/nllloss.hpp:239: Failure
> > > Value of: error < threshold * 10
> > >   Actual: false
> > > Expected: true
> > > Error output beyond tolerance Error:0.13130428906372638,  Thresholdx10: 0.078125
> > > 
> > > [  FAILED  ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12, where GetParam() =  input:{8192,52100} weight_mode:1 ignore_index:-100 divisor:1 Contiguous:0 (2132 ms)
> > > ```
> > 
> > 
> > My code passes the Driver test for this input even when I change the Tref to bfloat. The unit test passes for the values of the workspaces but still fails for the value of the reduce sum. I tried many ways, but it still doesn't work. The error remains at 0.13 (while the threshold is 0.07). Can we tolerate this?
> 
> Forgive me for being late.
> 
> In the kernel, bfloat16 is converted to float for calculation. Try the same as kernel.
> 
> ```
> root@a06c94249c27:/home/MIOpen/build# git diff
> diff --git a/test/cpu_nllloss.hpp b/test/cpu_nllloss.hpp
> index 87dff7338..7b5e258a9 100644
> --- a/test/cpu_nllloss.hpp
> +++ b/test/cpu_nllloss.hpp
> @@ -68,6 +68,7 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
>      auto I_tv = get_inner_expanded_tv_5d(input.desc);
>      auto T_tv = get_inner_expanded_tv_4d(target.desc);
>      auto W_tv = get_inner_expanded_tv_1d(weight.desc);
> +    float d   = !std::isnan(divisor) ? static_cast<float>(divisor) : static_cast<float>(1.0f);
> 
>      for(size_t i = 0; i < numel; i++)
>      {
> @@ -84,11 +85,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
>          }
>          else
>          {
> -            T w = weight[weight_index];
> -
> -            T input_value = input[input_index];
> -            T d           = !std::isnan(divisor) ? static_cast<T>(divisor) : static_cast<T>(1.0f);
> -            workspace[i]  = (-w * input_value) / d;
> +            float w = static_cast<float>(weight[weight_index]);
> +            float input_value = static_cast<float>(input[input_index]);
> +            workspace[i]  = static_cast<T>((-w * input_value) / d);
>          }
>      }
> 
> @@ -101,9 +100,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
>      {
>          for(int i = 0; i < _size; i += local_size)
>          {
> -            T shared[local_size];
> +            float shared[local_size];
>              par_ford(local_size)([&](size_t j) {
> -                shared[j] = i + j < _size ? workspace[offset_a + i + j] : static_cast<T>(0.0f);
> +               shared[j] = i + j < _size ? static_cast<float>(workspace[offset_a + i + j]) : static_cast<float>(0.0f);
>              });
>              for(int offset = local_size / 2; offset > 0; offset >>= 1)
>                  ford(local_size)([&](size_t j) {
> @@ -111,9 +110,9 @@ void cpu_nllloss_reduce_forward_5d(tensor<T> input,
>                          shared[j] += shared[j + offset];
>                  });
>              if(_size <= local_size)
> -                output[0] = shared[0];
> +                output[0] = static_cast<T>(shared[0]);
>              else
> -                workspace[offset_b + i / local_size] = shared[0];
> +                workspace[offset_b + i / local_size] = static_cast<T>(shared[0]);
>          }
>          std::swap(offset_a, offset_b);
>          _size = (_size + local_size - 1) / local_size;
> ```
> 
> ```
> # MIOPEN_TEST_FLOAT_ARG=--bfloat16 bin/test_nllloss
> 
> [----------] 13 tests from NLLLossTestSet/NLLLossTestBFloat16
> [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/0
> MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '"/root/.cache/miopen/3.1.0.56a22e6bf"' is: 0x794c7630 'OVERLAYFS_SUPER_MAGIC'
> MIOpen(HIP): Info [KernDb] database not present
> 
> ...
> 
> [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/11
> [       OK ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/11 (0 ms)
> [ RUN      ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12
> [       OK ] NLLLossTestSet/NLLLossTestBFloat16.NLLLossTest/12 (53164 ms)
> [----------] 13 tests from NLLLossTestSet/NLLLossTestBFloat16 (55032 ms total)
> ```

I see, fixed it in [commit](https://github.com/moreh-dev/MIOpen/pull/17/commits/fe750583174879fcfa36659e893eabfa611939f4)

**Comment by et16kr** [comment]
I mainly checked the parts that were different from PR 16.
You did a really good job.

**Pull request closed without merging**